clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -845,6 +882,30 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
         return parsedRecords;
     }
 
+    private PriorityQueue<FetchResponse.AbortedTransaction> getAbortedTransactions(FetchResponse.PartitionData partition) {
+        PriorityQueue<FetchResponse.AbortedTransaction> abortedTransactions = null;
+        if (partition.abortedTransactions != null && 0 < partition.abortedTransactions.size()) {
+            abortedTransactions = new PriorityQueue<>(
+                    partition.abortedTransactions.size(),
+                    new Comparator<FetchResponse.AbortedTransaction>() {
+                        @Override
+                        public int compare(FetchResponse.AbortedTransaction o1, FetchResponse.AbortedTransaction o2) {
+                            return (int) o1.firstOffset - (int) o2.firstOffset;
+                        }
+                    }
+            );
+            abortedTransactions.addAll(partition.abortedTransactions);
+        } else {
+            abortedTransactions = new PriorityQueue<>();
+        }
+        return abortedTransactions;
+    }
+
+    private boolean containsAbortMarker(RecordBatch batch) {
+        Record firstRecord = batch.iterator().hasNext() ? batch.iterator().next() : null;
```

Comment:
I guess more than one record in a control message is a bug. I will update the code to return a corrupted record exception in that case.